extensions/wikia/DesignSystem/scripts/DesignSystemGlobalNavigationOnSiteNotifications.js

```diff
@@ -1,42 +1,141 @@
+					OnSiteNotifications.markAllAsRead();
+				});
+			},
+
+			addDropdownLoadingEvent: function () {
+				var $dropdown = $('#on-site-notifications-dropdown');
+				$dropdown.click(function () {
+					OnSiteNotifications.loadFirstPage();
+				});
 			},
 
-			updateCounts: function () {
+			shouldLoadFirstPage: function () {
+				return !this.updateInProgress && this.nextPage && this.allPagesLoaded !== true;
+			},
+
+			shouldLoadNextPage: function () {
+				return !this.updateInProgress && !this.nextPage && this.allPagesLoaded !== true;
+			},
+
+			markAllAsRead: function () {
+				//TODO use the date of the first notification
+				this.bucky.timer.start('markAllAsRead');
+				$.ajax({
+					type: 'POST',
+					data: JSON.stringify({since: "2017-01-01T12:12:12.000Z"}),
+					dataType: 'json',
+					contentType: "application/json; charset=UTF-8",
+					url: this.getBaseUrl() + '/notifications/mark-all-as-read',
+					xhrFields: {
+						withCredentials: true
+					}
+				}).done(this.proxy(function () {
+					this.renderUnreadCount(0);
+					//TODO clear read status from all loaded notifications
+				})).always(this.proxy(function () {
+					this.bucky.timer.stop('markAllAsRead');
+				}));
+			},
+
+			loadFirstPage: function () {
+				if (this.shouldLoadFirstPage()) {
+					return;
+				}
+				this.updateInProgress = true;
+				this.bucky.timer.start('loadFirstPage');
+				$.ajax({
+					url: this.getBaseUrl() + '/notifications',
+					xhrFields: {
+						withCredentials: true
+					}
+				}).done(this.proxy(function (data) {
+					this.renderNotifications(this.mapToModel(data.notifications));
+					this.calculatePage(data);
+				})).always(this.proxy(function () {
+					this.updateInProgress = false;
+					this.bucky.timer.stop('loadFirstPage');
+				}));
+			},
+
+			calculatePage: function (data) {
+				this.nextPage = getSafely(data, '_links.next');
+				if (!this.nextPage) {
+					this.allPagesLoaded = true;
+				}
+			},
+
+			mapToModel: function (notifications) {
+				return notifications.map(function (notification) {
+					return {
+						title: getSafely(notification, 'refersTo.title'),
+						snippet: getSafely(notification, 'refersTo.snippet'),
+						uri: getSafely(notification, 'refersTo.uri'),
+						timestamp: getSafely(notification, 'events.latestEvent.when'),
+						communityName: getSafely(notification, 'community.name'),
+						communityId: getSafely(notification, 'community.id'),
+						isUnread: notification.read === false,
+						totalUniqueActors: getSafely(notification, 'events.totalUniqueActors')
+						// latestActors: NotificationModel.createActors(x.events.latestActors')),
+						// type: NotificationModel.getTypeFromApiData(notificationData)
+					};
+				});
+			},
+
+			renderNotifications: function (notifications) {
+				notifications.forEach(this.proxy(function (notification) {
+					var html = mustache.render(this.template, notification);
+					this.$container.append(html);
```

Comment:
I think it would be better to render all notifications and then append them to DOM at once. Also, what do you think about putting this loop inside the template and passing the array there? Then, the Mustache engine wouldn't need to be run multiple times.